src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkCustomCommandLine.java

```diff
@@ -858,6 +892,15 @@ private static String expandToCommandLine(
         : CommandLineItem.expandToCommandLine(object);
   }
 
+  private static void addSingleObjectToFingerprint(Fingerprint fingerprint, Object object) {
+    if (object instanceof FileApi) {
+      fingerprint.addUUID(FILE_TYPE_MARKER);
```

Comment:
I'm wondering if this call is going to add overhead (cpu and garbage) to fingerprinting lists of files. Is it safe to omit the marker if path mapping is disabled? That's one option if you can figure out how to get that information here.

Alternatively, maybe we could get away with adding fewer bits than a `UUID`, which is adding 128 bits to the fingerprint. We're just trying to distinguish among three possibilities, so consider adding booleans or ints (which are variable length encoded).

Reply:
As my preference is to keep the performance of path mapping as low as possible and visible to the standard benchmarks, I went with an attempt at the alternative solution first: I am now using an enum instead of a `UUID` and always add that to the fingerprint.

I kept the `instanceof` for now, but I could also extract this logic (and enum) into `CommandLineItem` if you prefer that. I don't know whether `instanceof` vs. regular polymorphism has any performance implications.

Reply:
I appreciate this. The `UUID` approach might have been fine, but we should be extra careful because of the history of path mapping changes harming performance when it's disabled.

I think the `instanceof` checks are acceptable.

Reply:
I also moved the fingerprinting of the enum into the `mapEach == null` branch.